hack/generator/pkg/codegen/pipeline_augment_status.go

```diff
@@ -0,0 +1,262 @@
+	"golang.org/x/sync/errgroup"
+	"k8s.io/klog/v2"
+)
+
+/* augmentResourcesWithStatus creates a PipelineStage to add status information into the generated resources.
+
+This information is derived from the Azure Swagger specifications. We parse the Swagger specs and look for
+any actions that appear to be ARM resources (have PUT methods with types we can use and appropriate names in the
+action path). Then for each resource, we use the existing JSON AST parser to extract the status type
+(the type-definition part of swagger is the same as JSON Schema).
+
+Next, we walk over all the resources we are currently generating CRDs for and attempt to locate
+a match for the resource in the status information we have parsed. If we locate a match, it is
+added to the Status field of the Resource type, after we have renamed all the status types to
+avoid any conflicts with existing Spec types that have already been defined.
+
+*/
+func augmentResourcesWithStatus(idFactory astmodel.IdentifierFactory, config *config.Configuration) PipelineStage {
+	return PipelineStage{
+		"augmentStatus",
+		"Add information from Swagger specs for 'status' fields",
+		func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
+
+			if config.Status.SchemaRoot == "" {
+				klog.Warningf("No status schema root specified, will not generate status types")
+				return types, nil
+			}
+
+			klog.V(1).Infof("Loading Swagger data from %q", config.Status.SchemaRoot)
+
+			swaggerTypes, err := loadSwaggerData(ctx, idFactory, config)
+			if err != nil {
+				return nil, errors.Wrapf(err, "unable to load Swagger data")
+			}
+
+			klog.V(1).Infof("Loaded Swagger data (%v resources, %v other types)", len(swaggerTypes.resources), len(swaggerTypes.otherTypes))
+
+			newTypes := make(astmodel.Types)
+
+			resourceLookup, otherTypes := makeStatusLookup(swaggerTypes)
+			for _, otherType := range otherTypes {
+				newTypes.Add(otherType)
+			}
+
+			found := 0
+			notFound := 0
+			for typeName, typeDef := range types {
+				if resource, ok := typeDef.Type().(*astmodel.ResourceType); ok {
+					if statusDef, ok := resourceLookup[typeName]; ok {
+						klog.V(4).Infof("Swagger information found for %v", typeName)
+						newTypes.Add(astmodel.MakeTypeDefinition(typeName, resource.WithStatus(statusDef)))
+						found++
+					} else {
+						// TODO: eventually this will be a warning
+						klog.V(2).Infof("No swagger information found for %v", typeName)
+						newTypes.Add(typeDef)
+						notFound++
+					}
+				} else {
+					newTypes.Add(typeDef)
+				}
+			}
+
+			klog.V(1).Infof("Found status information for %v resources", found)
+			klog.V(1).Infof("Missing status information for %v resources", notFound)
+			klog.V(1).Infof("Input %v types, output %v types", len(types), len(newTypes))
+
+			return newTypes, nil
+		},
+	}
+}
+
+type resourceLookup map[astmodel.TypeName]astmodel.Type
+
+// makeStatusLookup makes a map of old name (non-status name) to status type
```

Comment:
Comment (and possibly function name) seems out of date with signature -- there are two return types here but comment only talks about one. What is the other one?

Also, since this seemingly is making two things (and also altering some types in the process) unclear if `makeStatusLookup` is the best name.

Wondering about something like: `renameStatusDefinitions` or something?

Reply:
Renamed to `generateStatusTypes` and added clarifying comment.

Reply:
Reworked this greatly, actually. Should be more understandable now…